### PR TITLE
Convert async -> Result signature functions to async throws

### LINKED ISSUE #8104: Convert `await -> Result<T, Error>` functions to `await throws -> T`
I recently submitted a PR to [address a warning](https://github.com/woocommerce/woocommerce-ios/pull/8099). @jkmassel spotted the method I operated in has signature `await -> Result<Void, Error>`, at which he [commented](https://github.com/woocommerce/woocommerce-ios/pull/8099#pullrequestreview-1177924580) "the use of async -> Result is unfortunate here".

I think what Jeremy is hinting at is that, while evidently compiling correctly, `async` and `Result` are somewhat clashing from a functionality point of view.

## Background on `Result` and asycn/await

Prior to async/await, it was useful to use `Result` in asynchronous method to have a single callback that safely encapsulated both success and failure outcomes:

```swift
func someAsyncComputation(_ onCompletion: Result<Int, Error>)
```

With async/await, we have an alternative pattern available

```swift
func someComputation() async throws -> Int
```

Of course, the option used in the codebase also compiles, but it feels a bit stuck in the middle.

```swift
func someComputation() async -> Result<Int, Error>
```

Here's the difference at call site:

```swift
// Result in closure
someAsyncComputation { result in 
  switch result {
  case .success(let value): ...
  case .failure(let error): ...
  }
}

// async throws
do {
  let value = try await someComputation()
  // ...
} catch {
  // ...
}

// async throws in a function that itself throws to bubble up the error
let value = try await someComputation

// async Result
let result = await someComputation()
switch result  {
  case .success(let value): ...
  case .failure(let error): ...
}
```

Personally, I am a big fan of `Result` and I love seeing it in use. But I think it works best when chaining functional programming style transformations in [railway oriented programming](https://fsharpforfunandprofit.com/rop/):

```swift
result
  .map { ... }
  .flatMap { ... }
  .mapError { ... }
  // etc
```

## Where to go from here

At the time of writing, there are 24 matches for `async -> Result` in the codebase. 

image

I sampled them and I don't see us leveraging `Result` in a railway oriented programming way. Example see `loadUsernameSuggestions(from:)` defined as

https://github.com/woocommerce/woocommerce-ios/blob/d0a66f4e85ec7eaf54258f7ac6b22247adb6d2d5/Networking/Networking/Remote/AccountRemote.swift#L16

and used as

https://github.com/woocommerce/woocommerce-ios/blob/d0a66f4e85ec7eaf54258f7ac6b22247adb6d2d5/Yosemite/Yosemite/Stores/AccountCreationStore.swift#L80-L85

We could rewrite this to something like

```swift
return try? await remote.loadUsernameSuggestions(from: base)?.first
```

It's also possible we'll find some calls that ought to remain callback-based and could benefit with `Result` without `async`. 

---

I'd love to hear your thoughts on this. Thanks 😄

### PR DESCRIPTION
**Pull request by toupper**
Closes: #8104 (Sorry for the bigger PR)

### Description

With this PR I convert the function signatures with `await -> Result<..., Error>` to the most natural `async throws -> ...`, except in those cases where we have a specific error type e.g [here](https://github.com/woocommerce/woocommerce-ios/blob/636a27307afac5c2cc1ac7cb45dc349172c08c23/WooCommerce/Classes/Authentication/Store%20Creation/StoreCreationCoordinator.swift#L220)

### Testing instructions

Unit tests should pass. You can also check that the related features do not present any side effects:
- Remove Apple Id access
- Create account
- Domain suggestions
- Username suggestions
- Free domain suggestions
- Site creation
- Just in Time Messages
- Stats

### Screenshots

---
- [X] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Networking/Networking/Remote/Remote.swift
Comment (koke) [inline Networking/Networking/Remote/Remote.swift:215]: I was going to suggest that we could replace this chunk of code using the existing helpers to convert a throwing call to a Result, and a Result to a continuation. After trying it, the code is similarly long and I believe less readable, so the current approach feels better 🤷🏽‍♂️ 

Leaving it here in case it helps for other future implementations:

```swift
continuation.resume(
    with: result.flatMap { data in
        Result {
            let validator = request.responseDataValidator()
            try validator.validate(data: data)
            return try mapper.map(response: data)
        }
    }.mapError({ error in
        DDLogError("<> Mapping Error: \(error)")
        self.handleResponseError(error: error, for: request)
        return error
    })
)
```

**Comment by jaclync** [comment]
Thanks for tagging me in this one! I see that this PR is going to conflict with another PR of mine https://github.com/woocommerce/woocommerce-ios/pull/8154 which is ready to merge but is blocked by CI, I just disabled auto-merge in case of bad timing. I'm also working on another task that will conflict with the part about removing Apple ID access https://github.com/woocommerce/woocommerce-ios/issues/8178, I will follow this PR and rebase when needed.

Updated: the second PR with potential merge conflicts is https://github.com/woocommerce/woocommerce-ios/pull/8179

File: WooCommerce/Classes/ViewRelated/Authentication/AccountCreationForm.swift
Comment (koke) [inline WooCommerce/Classes/ViewRelated/Authentication/AccountCreationForm.swift]: Perhaps we can use `try?` here to simplify this block if we want to ignore the error case

File: Yosemite/Yosemite/Stores/DomainStore.swift
Comment (koke) [inline Yosemite/Yosemite/Stores/DomainStore.swift]: I was going to suggest `Result(catching:)` here but then realized it's async. In the original conversation, I suggested an extension to make an async catching initializer:

```swift
extension Result where Failure == Error {
    init(catching body: () async throws -> Success) async {
        do {
            let value = try await body()
            self = .success(value)
        } catch {
            self = .failure(error)
        }
    }
}
```

With that, this could be simplified as:

```swift
private extension DomainStore {
    func loadFreeDomainSuggestions(query: String, completion: @escaping (Result<[FreeDomainSuggestion], Error>) -> Void) {
        Task { @MainActor in
            let result = await Result {
                try await remote.loadFreeDomainSuggestions(query: query)
            }
            completion(result)
        }
    }
}
```

I see this also apply to several other diffs in Yosemite, so it might make sense to add it on this PR

File: Yosemite/YosemiteTests/Mocks/Networking/Remote/MockAccountRemote.swift
Comment (koke) [inline Yosemite/YosemiteTests/Mocks/Networking/Remote/MockAccountRemote.swift]: `Result.get()` is a convenient way to convert from result to throwing:

```suggestion
        return try result.get()
```

I see a couple other instances of this in YosemiteTests

**Review by koke — APPROVED** [review]
Made a few suggestions to simplify a bit, but overall it looks good. When I started reviewing I saw unit tests failing, but whenever those are fixed I'd say it's good.

File: WooCommerce/Classes/ViewRelated/Authentication/AccountCreationForm.swift
Comment (toupper) [reply WooCommerce/Classes/ViewRelated/Authentication/AccountCreationForm.swift]: Sure, done in 5b2efe2b3cda20c88dc945a3ab2ff2fb27740f3f

File: Yosemite/YosemiteTests/Mocks/Networking/Remote/MockAccountRemote.swift
Comment (toupper) [reply Yosemite/YosemiteTests/Mocks/Networking/Remote/MockAccountRemote.swift]: True, done in 1a62a27e3ad8a9a3c5c42fa9f8786d28f026c6bb

File: Yosemite/Yosemite/Stores/DomainStore.swift
Comment (toupper) [reply Yosemite/Yosemite/Stores/DomainStore.swift]: Nice improvement! Added in b00cd6aa3b8590bee3b98455ed680b2385f12c10

**Comment by toupper** [comment]
@jaclync Thanks for letting me know about these possible conflicts! Do you want me to wait until your PRs are merged before merging this one?

**Comment by jaclync** [comment]
> Do you want me to wait until your PRs are merged before merging this one?

It depends on when the CI is fixed 😅 I was hoping to merge https://github.com/woocommerce/woocommerce-ios/pull/8154 ASAP since multiple branches (one PR also ready to merge) depend on it and I'll resolve the merge conflicts in your branch here, but the CI is still not working ATM and I don't know if will by the EOD my time today.

To unblock other work that depends on https://github.com/woocommerce/woocommerce-ios/pull/8154 and your branch, I'm going to merge your branch to https://github.com/woocommerce/woocommerce-ios/pull/8154 branch locally and resolve merge conflicts. If the CI is fixed during my day, I'll merge https://github.com/woocommerce/woocommerce-ios/pull/8154 and then resolve merge conflicts in a commit based on your branch (you can decide whether to use it). Otherwise, feel free to merge your PR if CI is ready during your day and I'll resolve the conflicts tomorrow 🙂

**Comment by jaclync** [comment]
Update: actually, there are no merge conflicts with the main PR I was hoping to merge! I probably misread the changes earlier. Another PR https://github.com/woocommerce/woocommerce-ios/pull/8179 is likely having merge conflicts but this is at a lower priority and I can handle them later. Please go ahead with merging this anytime! 😄

**Comment by toupper** [comment]
Great, thanks for the update @jaclync !

**Review by joshheald — APPROVED** [review]
Thanks for tagging me in this. I don't have any comments to add to Koke's and Jaclyn's, but the code looks good to me and tests well. I always appreciate learning from a PR!

**Comment by mokagio** [comment]
💜

**Pull request merged**
